### PR TITLE
fdtdump and FDT writer

### PR DESCRIPTION
**Pull request by rjoleary**

### CODE REVIEW

File: pkg/dt/fdt.go [hunk 1]
```diff
@@ -279,26 +283,130 @@ func (fdt *FDT) readStructBlock(f io.ReadSeeker, strs []byte) error {
...
+			if v != 0 {
+				// TODO: Some of the padding is not zero. Is this a mistake?
+				//return fmt.Errorf("padding is non-zero: %d", v)
 			}
 		}
 	}
 }
 
+func align(x, n int) uint32 {
+	return uint32((x + n - 1) / n * n)
```
[the remark this answers is not in the document]
Comment (rjoleary) [reply pkg/dt/fdt.go]: I changed it to an `&^`.

File: pkg/dt/fdt.go [hunk 2]
```diff
@@ -211,15 +218,14 @@ func (fdt *FDT) readStructBlock(f io.ReadSeeker, strs []byte) error {
 			// The name is a null-terminating string.
 			for {
 				c := make([]byte, 1)
```
Comment (n-canter) [inline pkg/dt/fdt.go]: I think it'd be nice if AlignReader implements ReadByte method (io.ByteReader interface)

File: pkg/dt/fdt.go [hunk 3]
```diff
@@ -279,26 +283,134 @@ func (fdt *FDT) readStructBlock(f io.ReadSeeker, strs []byte) error {
...
+	}
+	if err := writeNode(fdt.RootNode); err != nil {
+		return w.N, err
+	}
+	if err := binary.Write(w, binary.BigEndian, tokenEnd); err != nil {
+		return w.N, err
+	}
+
+	// Write strings block
+	if _, err := w.Write(strs); err != nil {
```
Comment (n-canter) [inline pkg/dt/fdt.go]:
```
err = w.Write(strs)
return w.N, err
```

File: pkg/dt/fdt.go (hunk 2 above)
Comment (rjoleary) [reply pkg/dt/fdt.go]: Done!

File: pkg/dt/fdt.go (hunk 3 above)
Comment (rjoleary) [reply pkg/dt/fdt.go]: Thanks!

File: pkg/dt/fdt.go [hunk 4]
```diff
@@ -279,26 +282,130 @@ func (fdt *FDT) readStructBlock(f io.ReadSeeker, strs []byte) error {
...
+	}
+	if err := writeNode(fdt.RootNode); err != nil {
+		return w.N, err
+	}
+	if err := binary.Write(w, binary.BigEndian, tokenEnd); err != nil {
+		return w.N, err
+	}
+
+	// Write strings block
+	return w.Write(strs)
```
Comment (n-canter) [inline pkg/dt/fdt.go]: I think you should return a size of all writes, not the size of the last Write().

**Pull request merged**